compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs

```diff
@@ -1259,6 +1270,22 @@ impl<'a, 'tcx> BoundVarContext<'a, 'tcx> {
         if let Some(mut def) = result {
             if let ResolvedArg::EarlyBound(..) = def {
                 // Do not free early-bound regions, only late-bound ones.
+            } else if let ResolvedArg::LateBound(_, _, param_def_id) = def
```

Comment:
Both this branch and the next (`if let Some(body_id) = outermost_body`) should only be executed for late-bound variables. Should they both be in an `if let ResolvedArg::LateBound(_, _, param_def_id) = def` to make it clearer?

For my curiosity, what would happen if we made lifetimes in anon-consts `ResolvedArg::Free`?

Reply:
> For my curiosity [...]

See my comment here https://github.com/rust-lang/rust/pull/115486#discussion_r1314015308, I kinda went into it there -- this idea would, in fact, fix at least tests like `const-generics/late-bound-vars/in_closure.rs` and `const-generics/late-bound-vars/simple.rs`, and was how #79298 worked, afaict.

It doesn't seem like a general solution, though, since when it comes to, e.g., a binder on a where clause like `where for<'a> [(); { /* const that uses 'a */ }]: ...`, it doesn't really make sense to use `ReFree` for that usage of `'a`, since the late-bound region does not originate from the function's generics (and therefore the `fn_sig`'s binder).

This also doesn't lead us towards a correct solution for handling things like `for<const C: usize> [(); C]: ...`, since there's no equivalent notion of a "free" type or const var (nor does it make sense, since they must be substitutable).